File: autosplitter.lua
obs = obslua

local enabled = false
local interval = 60

function script_description()
    return "AutoSplitter\nSplit recordings at a specific interval"
end

function getInterval(settings)
    local seconds = obs.obs_data_get_int(settings, "interval_s")
    local minutes = obs.obs_data_get_int(settings, "interval_m")
    local hours = obs.obs_data_get_int(settings, "interval_h")
    return hours, minutes, seconds
end

function setInterval(settings, hours, minutes, seconds)
    obs.obs_data_set_int(settings, "interval_s", seconds)
    obs.obs_data_set_int(settings, "interval_m", minutes)
    obs.obs_data_set_int(settings, "interval_h", hours)
end

function toSeconds(hours, minutes, seconds)
    return seconds + (minutes + hours * 60) * 60
end

function fromSeconds(interval)
    local seconds = interval % 60
    local minutes = math.floor(interval / 60) % 60
    local hours = math.floor(math.floor(interval / 60 / 60))
    return hours, minutes, seconds
end

local cancelCheck = function()
end
local cancelRestart = function()
end

function script_update(settings)
    enabled = obs.obs_data_get_bool(settings, "enabled")
    interval = toSeconds(getInterval(settings))
    if interval <= 0 then
        interval = 1
    end

    cancelCheck()
    cancelRestart()

    if enabled then
        cancelCheck = onRecordingChanged(
                function(recording)
                    if recording then
                        cancelRestart = timer(recording_restart, interval * 1000)
                    else
                        cancelRestart()
                    end
                end
        )
    end
end

function script_defaults(settings)
    obs.obs_data_set_default_bool(settings, "enabled", enabled)

    local hours, minutes, seconds = fromSeconds(interval)
    obs.obs_data_set_default_int(settings, "interval_s", seconds)
    obs.obs_data_set_default_int(settings, "interval_m", minutes)
    obs.obs_data_set_default_int(settings, "interval_h", hours)
end

function script_properties()
    local props = obs.obs_properties_create()

    obs.obs_properties_add_bool(props, "enabled", "Enabled")
    local prop_interval_s = obs.obs_properties_add_int(props, "interval_s", "Seconds", -1, 60, 1)
    local prop_interval_m = obs.obs_properties_add_int(props, "interval_m", "Minutes", -1, 60, 1)
    local prop_interval_h = obs.obs_properties_add_int(props, "interval_h", "Hours", 0, 240, 1)

    function validate(props, prop, settings)
        local hours, minutes, seconds = getInterval(settings)
        local interval = toSeconds(hours, minutes, seconds)
        if interval <= 0 then
            interval = 1
        end
        local newHours, newMinutes, newSecods = fromSeconds(interval)

        if hours == newHours and
                minutes == newMinutes and
                seconds == newSecods then
            return false
        else
            setInterval(settings, newHours, newMinutes, newSecods)
            return true
        end
    end

    obs.obs_property_set_modified_callback(prop_interval_s, validate)
    obs.obs_property_set_modified_callback(prop_interval_m, validate)
    obs.obs_property_set_modified_callback(prop_interval_h, validate)

    return props
end

function timer(func, millis)
    local cancelled = false

    obs.timer_add(
            function()
                if not cancelled then
                    func()
                else
                    obs.remove_current_callback()
                end
            end,
            millis
    )

    return function()
        cancelled = true
    end
end

function delay(func, millis)
    return timer(
            function()
                obs.remove_current_callback()
                func()
            end,
            millis
    )
end

function delayUntil(func, condition, millis)
    local cancel
    cancel = timer(
            function()
                if condition() then
                    cancel()
                    func()
                end
            end,
            millis
    )

    return cancel
end

function timerWhile(func, condition, millis)
    local cancel
    cancel = timer(
            function()
                if condition() then
                    func()
                else
                    cancel()
                end
            end,
            millis
    )

    return cancel
end

function recording_stopped(func)
    delayUntil(
            func,
            function()
                return not obs.obs_frontend_recording_active()
            end,
            10
    )
end

function recording_restart()
    if obs.obs_frontend_recording_active() and not obs.obs_frontend_recording_paused() then
        obs.obs_frontend_recording_stop()

        recording_stopped(
                function()
                    timerWhile(
                            function()
                                obs.obs_frontend_recording_start()
                            end,
                            function()
                                return not obs.obs_frontend_recording_active()
                            end,
                            10
                    )
                end
        )
    end
end

function onRecordingChanged(func)
    local recording = not obs.obs_frontend_recording_active()

    return timer(
            function()
                if recording ~= obs.obs_frontend_recording_active() then
                    recording = not recording
                    func(recording)
                end
            end,
            200
    )
end

File: AutoSplitter.py
import obspython as obs

enabled = False
interval = 60


def script_description():
    return "Restarts recording"


def script_update(settings):
    global enabled
    global interval

    enabled = obs.obs_data_get_bool(settings, "enabled")

    interval = get_interval(settings)
    if interval == 0:
        interval = 1

    def doRestart():
        if enabled:
            recording_restart()

    def onStart():
        obs.timer_remove(doRestart)
        if enabled:
            obs.timer_add(doRestart, interval * 1000)

    def onStop():
        obs.timer_remove(doRestart)

    check_recording(onStart, onStop)


def script_defaults(settings):
    global enabled
    global interval

    obs.obs_data_set_default_bool(settings, "enabled", enabled)
    obs.obs_data_set_default_int(settings, "interval_s", interval % 60)
    obs.obs_data_set_default_int(settings, "interval_m", int(interval / 60))
    obs.obs_data_set_default_int(settings, "interval_h", int(int(interval / 60) / 60))


def script_properties():
    props = obs.obs_properties_create()

    obs.obs_properties_add_bool(props, "enabled", "Enabled")
    prop_interval_s = obs.obs_properties_add_int(props, "interval_s", "Seconds", 0, 59, 1)
    prop_interval_m = obs.obs_properties_add_int(props, "interval_m", "Minutes", 0, 59, 1)
    prop_interval_h = obs.obs_properties_add_int(props, "interval_h", "Hours", 0, 240, 1)

    def validate(props, prop, settings):
        if get_interval(settings) == 0:
            obs.obs_data_set_int(settings, "interval_s", 1)
            return True
        else:
            return False

    obs.obs_property_set_modified_callback(prop_interval_s, validate)
    obs.obs_property_set_modified_callback(prop_interval_m, validate)
    obs.obs_property_set_modified_callback(prop_interval_h, validate)

    return props


# ------------------------------------------------------------

def runAfter(millis, callback):
    def do():
        obs.remove_current_callback()
        callback()

    obs.timer_add(do, millis)


def runWhen(interval, condition, callback):
    def check():
        if condition():
            obs.remove_current_callback()
            callback()

    obs.timer_add(check, interval)


def runWhile(interval, condition, callback):
    def check():
        if condition():
            callback()
        else:
            obs.remove_current_callback()

    obs.timer_add(check, interval)


def recording_stopped(callback):
    runWhen(10, lambda: not obs.obs_frontend_recording_active(), callback)


def recording_restart():
    if obs.obs_frontend_recording_active():
        obs.obs_frontend_recording_stop()
        recording_stopped(lambda: runWhile(
            10,
            lambda: not obs.obs_frontend_recording_active(),
            lambda: obs.obs_frontend_recording_start()
        ))


def check_recording(onStart, onStop):
    def check_active():
        if obs.obs_frontend_recording_active():
            obs.remove_current_callback()
            obs.timer_add(check_inactive, 200)
            onStart()

    def check_inactive():
        if not obs.obs_frontend_recording_active():
            obs.remove_current_callback()
            obs.timer_add(check_active, 200)
            onStop()

    obs.timer_remove(check_active)
    obs.timer_remove(check_inactive)
    obs.timer_add(check_active, 200)


def get_interval(settings):
    interval_s = obs.obs_data_get_int(settings, "interval_s")
    interval_m = obs.obs_data_get_int(settings, "interval_m")
    interval_h = obs.obs_data_get_int(settings, "interval_h")
    return interval_s + (interval_m + interval_h * 60) * 60

# ------------------------------------------------------------
